web/api/v1/api.go

```diff
@@ -1746,12 +1791,60 @@ OUTER:
 	return matcherSets, nil
 }
 
+// marshalPointJSON writes `[ts, "val"]`.
 func marshalPointJSON(ptr unsafe.Pointer, stream *jsoniter.Stream) {
 	p := *((*promql.Point)(ptr))
 	stream.WriteArrayStart()
+	marshalTimestamp(p.T, stream)
+	stream.WriteMore()
+	marshalValue(p.V, stream)
+	stream.WriteArrayEnd()
+}
+
+func marshalPointJSONIsEmpty(ptr unsafe.Pointer) bool {
+	return false
+}
+
+// marshalExemplarJSON writes.
+// {
+//    labels: <labels>,
+//    value: "<string>",
+//    timestamp: <float>
+// }
```

Comment:
Can I ask why `hasTs` field is not included here? It would be useful for Thanos's proxy implementation.

Reply:
Thanos proxy gets exemplar from Prometheus by calling `/query_exemplars` API from the sidecar. If `hasTs` field is not included in the API, then is there a way to tell whether this exemplar has a timestamp or not? 

IIUC if the exemplar doesn't have a ts, the timestamp is 0? @cstyan

Reply:
HasTs is an internal implementation detail. From the apy you will always get a timestamp.

Reply:
Yeah as Julien said `HasTs` is an internal implementation detail. An exemplar can optionally have a timestamp when it is exported, the same as a regular sample. If the exemplar didn't have it's own timestamp set, `HasTs` will be false and internal storage will assign the scrape timestamp as the exemplars timestamp. Any query to Prometheus for exemplars will always return exemplars with timestamps.

Reply:
Thanks, it is clear now.